os_window_x11.c

```diff
@@ -309,29 +368,42 @@ void
 os_window_sync_buffer_to_screen(struct os_window_struct* p_window) {
   int ret;
 
-  Bool bool_ret = XShmPutImage(p_window->d,
-                               p_window->w,
-                               p_window->gc,
-                               p_window->p_image,
-                               0,
-                               0,
-                               0,
-                               0,
-                               p_window->width,
-                               p_window->height,
-                               False);
-  if (bool_ret != True) {
-    errx(1, "XShmPutImage failed");
-  }
-
-  /* We need to sync here so that the server ack's it has finished the
-   * XShmPutImage.
-   * Clients of this function expect to be able to start writing to the buffer
-   * again immediately without affecting display.
-   */
-  ret = XSync(p_window->d, False);
-  if (ret != 1) {
-    errx(1, "XSync failed");
+  if (p_window->use_mit_shm) {
+    Bool bool_ret = XShmPutImage(p_window->d,
+                                 p_window->w,
+                                 p_window->gc,
+                                 p_window->p_image,
+                                 0,
+                                 0,
+                                 0,
+                                 0,
+                                 p_window->width,
+                                 p_window->height,
+                                 False);
+    if (bool_ret != True) {
+      errx(1, "XShmPutImage failed");
+    }
+    
+    /* We need to sync here so that the server ack's it has finished the
+     * XShmPutImage. Clients of this function expect to be able to start writing
+     * to the buffer again immediately without affecting display.
+     */
+    ret = XSync(p_window->d, False);
```

Comment:
I think this XSync check needs to occur in both the SHM and non-SHM case? i.e. put it after the parent if/else? (This may slow down a slow situation further as it will stop the beebjit thread and the X thread from doing stuff in parallel. But there's -opt video:no-vsync-wait-for-render for precisely this case.)